jaxopt/_src/eq_qp.py

```diff
@@ -57,7 +57,8 @@ def eq_fun(primal_var, params_eq):
 
   # It is required to post_process the output of `idf.make_kkt_optimality_fun`
   # to make the signatures of optimality_fun() and run() agree.
-  def optimality_fun(params, params_obj, params_eq):
+  # The M argument is needed for using preconditioners.
+  def optimality_fun(params, params_obj, params_eq, M=None):
```

Comment:
On first sight, I'm rather -1 on introducing a pre-conditioner `M` in `optimality_fun` since I don't think it plays any role in the optimality conditions (it would not make sense to differentiate with respect to `M`). Since `run` and `optimality_fun` need to have the same signature, this rules out adding `M` to `run` as well.

I think I would go for something like this instead:
```python
preconditioner = PseudoInversePreconditioner(params_obj, params_eq)
qp = EqualityConstrainedQP(preconditioner=preconditioner)
qp.run(params_obj=params_obj, params_eq=params_eq)
```
Typically, stuff that doesn't need to be differentiated should go to the constructor.

If you want to differentiate wrt `params_eq` or `params_obj`, you may need to use 
```python
EqualityConstrainedQP(preconditioner=lax.stop_gradient(preconditioner))
```
instead. Not entirely sure if `PseudoInversePreconditioner` should live in JAXopt or in user land.

Reply:
It doesn't play a role in the optimality conditions -- it could play a role in the backwards though, if we manage to pass the argument to the backward solver as well. This would hopefully speed things up, since the forward and backward linear systems share the linear operator.

With the current API, if we're solving the QP as the inner problem of a bi-level problem, then we need to build a new QP solver instance at each step of the outer loop, and pass solve=partial(linearsolver, M=preconditioner) to both `solve` and `implicit_diff_solve`.

Something which is then unclear to me: does building a new instance of the QP solver necessarily trigger recompilation of the `run` method at each iteration ?